File: src/CPP/code/ctci/ch_1_Arrays_and_Strings/StringCompression.cpp
#pragma clang diagnostic push
#pragma ide diagnostic ignored "LocalValueEscapesScope"

#include "StringCompression.h"
#include <iostream>
#include <string>

using namespace ctci;

void StringCompression::description() {
    std::cout << R"(Arrays and Strings: String Compression
    Implement a method to perform basic string compression using the counts of repeated characters.
    For example, the string aabcccccaaa would become a2blc5a3.
    If the "compressed" string would not become smaller than the original string, your method should return
    the original string.
    You can assume the string has only uppercase and lowercase letters (a - z).)" << std::endl;
}

std::string StringCompression::compress(const std::string &input) {
    std::string result;
    auto count = 0;
    char last = input.at(0);
    for (auto character: input) {
        if (character != last) {
            result.push_back(last);
            result.append(std::to_string(count));
            last = character;
            count = 1;
        } else {
            count++;
        }
    }
    // add last character count to result
    result.push_back(last);
    result.append(std::to_string(count));
    return (result.size() < input.size()) ? result : input;
}

std::string StringCompression::decompress(const std::string &input) {
    std::string result;
    char last;
    for (auto index = 0; index < input.size(); index++) {
        if (isdigit(input.at(index))) {
            auto start = index;
            auto count = 1;
            while ((index < input.size()) and (isdigit(input.at(index)))) {
                count++;
                index++;
            }
            // index is pointing to the next character but the loop will increment again, so I need to decrement
            index--;
            count--;
            std::string number = input.substr(start, count);
            auto append = std::string(stoi(number), last);
            result.append(append);
        } else {
            last = input.at(index);
        }
    }
    return (result.empty()) ? input : result;
}

#pragma clang diagnostic pop

File: src/Python/Code/CtCI/ch_1_Arrays_and_Strings/StringCompression.py
class StringCompression(object):
    @staticmethod
    def get_description():
        print("""Arrays and Strings: String Compression
    Implement a method to perform basic string compression using the counts of repeated characters.
    For example, the string aabcccccaaa would become a2blc5a3.
    If the "compressed" string would not become smaller than the original string, your method should return
    the original string.
    You can assume the string has only uppercase and lowercase letters (a - z).""")

    @staticmethod
    def compress(base: str) -> str:
        if len(base) == 0:
            return base
        values = []
        char_list = list(base)
        last = ''
        counter = 0
        while char_list:
            if char_list[0] == last:
                counter += 1
                char_list.pop(0)
            else:
                if counter != 0:
                    values.append(str(last) + str(counter))
                last = char_list.pop(0)
                counter = 1
        if counter != 0:
            values.append(str(last) + str(counter))
        return min(base, "".join(values), key=len)

    @staticmethod
    def decompress(base: str) -> str:
        if len(base) == 0:
            return base
        values = []
        char_list = list(base)
        last = ''
        while char_list:
            current = char_list.pop(0)
            if current.isdigit():
                append = ''
                while char_list and char_list[0].isdigit():
                    current += char_list.pop(0)
                for i in range(int(current)):
                    append += last
                values.append(append)
                last = ''
            elif last == '':
                last = current
            else:
                values.append(last)
                last = current
        if last != '':
            values.append(last)
        return "".join(values)
